### ISSUE TITLE
java.lang.RuntimeException: Cloud size under 11 - again (on Spark 2.2, Sparkling water 2.2.2)

### ISSUE THREAD

**Issue by ibobak** [h2oai/sparkling-water#433]
I reported my problem in https://github.com/h2oai/sparkling-water/issues/32  but then I saw that it was closed.  If you don't mind, I would like to start a new issue, since the problem reproduces with Spark 2.2 and Sparkling water 2.2.2.

Here are my configs in the Notebook's kernel:

```
"PYSPARK_SUBMIT_ARGS":" --py-files /usr/local/share/jupyter/kernels/sparkling-water-2.2.2/py/build/dist/h2o_pysparkling_2.2-2.2.2.zip 
  --conf \"spark.scheduler.minRegisteredResourcesRatio=1\" 
  --conf \"spark.ext.h2o.topology.change.listener.enabled=false\" 
  --conf \"spark.ext.h2o.fail.on.unsupported.spark.param=false\" 
  --conf \"spark.ext.h2o.node.network.mask=10.5.33.0/24\" 
  --jars /usr/local/share/jupyter/kernels/aws-lib/hadoop-aws-2.7.3.jar,/usr/local/share/jupyter/kernels/aws-lib/aws-java-sdk-1.7.4.jar   
  --driver-memory 8G 
  --executor-memory 24G   
  --conf \"spark.dynamicAllocation.enabled=false\" 
  --num-executors 10 
  --executor-cores 2 
  --master spark://10.5.33.36:7077 pyspark-shell"   
```
  
No matter of applying all these settings, I am receiving the same error:

```
Py4JJavaError: An error occurred while calling z:org.apache.spark.h2o.JavaH2OContext.getOrCreate.
: java.lang.RuntimeException: Cloud size under 11
	at water.H2O.waitForCloudSize(H2O.java:1689)
	at org.apache.spark.h2o.backends.internal.InternalH2OBackend.init(InternalH2OBackend.scala:117)
	at org.apache.spark.h2o.H2OContext.init(H2OContext.scala:121)
	at org.apache.spark.h2o.H2OContext$.getOrCreate(H2OContext.scala:355)
	at org.apache.spark.h2o.H2OContext$.getOrCreate(H2OContext.scala:371)
	at org.apache.spark.h2o.H2OContext.getOrCreate(H2OContext.scala)
	at org.apache.spark.h2o.JavaH2OContext.getOrCreate(JavaH2OContext.java:228)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:280)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:214)
	at java.lang.Thread.run(Thread.java:748)
```
 
I the Spark app I clearly see that it started one driver and 10 executors, and (as you see) the amount of executors is explicitly configured. No matter of that, this annoying error simply doesn't allow to run the H2O.

I'll be very grateful for any ideas of how to run it.

**Comment by jakubhava** [comment]
Hi @ibobak, this is  a bit different issue. Please see a few links bellow which I think are a good starting point to understand what is happening there.

I would suggest reading about the backends here https://github.com/h2oai/sparkling-water/blob/master/doc/tutorials/backends.rst and if you need to stick with the internal backend the tuning guide https://github.com/h2oai/sparkling-water/blob/master/doc/configuration/internal_backend_tuning.rst.

One important thing which helps to solve this issue is disabling Spark dynamic allocation `spark.dynamicAllocation.disabled` as then we can discover the executors more deterministically.

However we are not able to get rid of this problem completely in the internal sparkling water backend. In order to be able to start h2o on all executors we need to first find all the spark executors. However spark doesn't give us this information right away and thus we try to utilize the whole cluster to discover all executors. This PR might https://github.com/h2oai/sparkling-water/pull/297 help as we might be running just on portion of spark executors.

**Comment by ibobak** [comment]
One thing is not clear in the documentation about external backends. It is said:
```
To start external H2O cluster where the nodes are discovered using the flatfile, you can run
java -jar $H2O_EXTENDED_JAR -md5skip -name test -flatfile path_to_flatfile
```

Does it mean that I need to run this ONCE at any node?  I am just curious, how it will connect to those machines and how it will run processes?  Will it ssh to those nodes, copy there jar files and run the processes on them?  Under what account it will do? It would be nice if you tell some words about the things that would be happening if I do this.

Besides, where is the parameters of the amount of memory and the number of cores which will be assigned to the H2O nodes? I think it is the most basic thing, although it is missing.

**Comment by jakubhava** [comment]
There are two modes of the external backend - manual and automatic.
The automatic mode is available on yarn only and will take care of starting the external H2O cluster automatically. In case of manual mode you have 2 options ( but the documentation mention just 1 )

- you either really need to start the h2o cluster using the script above on all the nodes where you want your cluster to be running. To see all available h2o cli options, please see http://h2o-release.s3.amazonaws.com/h2o/rel-weierstrass/7/docs-website/h2o-docs/starting-h2o.html?highlight=starting%20h2o. H2O clouds up either using multicast or flat-file. Please see the H2O documentation where this is explained and reach here if you have more questions.
- if you have yarn, you can start H2O on hadoop. Please this H2O documentation http://h2o-release.s3.amazonaws.com/h2o/rel-weierstrass/7/docs-website/h2o-docs/downloading.html?highlight=hadoop. This is how is usually H2O deployed in production. The reason why this is not mentioned in the documentation is the assumption that when people have YARN available, we to bother them with manual mode if we can do it automatically.

**Comment by jakubhava** [comment]
So the outcome is - if you have YARN available and want to use external cluster, I really suggest using automatic deployment mode. Here https://github.com/h2oai/sparkling-water/blob/master/doc/configuration/configuration_properties.rst#external-backend-configuration-properties you can find the configuration properties you can use in sparkling water to affect formation of external h2o cluster in automatic mode. They are just simple wrappers around h2o configuration.

**Comment by WeidongZhou** [comment]
Hi, @ibobak,
If I remember correctly, I saw your other comment about the similar issues about 10 executors on BDA no matter what parameters you gave. I actually just fixed similar issue for one of my clients also running their stuff on a full rack BDA as well. The issue was on R Studio and never get anything more than 10 executors. After investigation, I found out this 10 executors is the default value. If you check closely, you will find one executor will use only one core. The main issue in my client is the way they pass parameters was wrong. So the issue in your case is very likely the way you pass parameters. Try pass parameters for executors like the following:
--conf spark.driver.memory=8G
--conf spark.executor.cores=2
--conf spark.executor.memory=24G
--conf spark.executor.instances=10

Hope, it works for you.

Weidong

**Comment by chrismoran0124** [comment]
I was having the same issue trying to run Spark 2.2, Sparkling Water 2.2.2, and sparklyr within R Studio. I tried it a bunch of different ways using different configurations suggested in this and previous threads. Having the right versions I believe helped: Spark 2.2.0, Sparkling Water 2.2.2 and h20 version 3.14.07
I got a suggestion in R Studio when connecting to Spark:

```
Detected H2O version 3.14.0.7. Please install H2O version 3.14.0.2, which is compliant with the latest Sparkling Water version for Spark 2.2.* -> Sparkling Water version 2.2.2

                       To update your h2o R package, copy/paste the following commands and then restart your R session:
                       
                       detach("package:rsparkling", unload = TRUE)
                       if ("package:h2o" %in% search()) { detach("package:h2o", unload = TRUE) }
                       if (isNamespaceLoaded("h2o")){ unloadNamespace("h2o") }
                       remove.packages("h2o")
                       install.packages("h2o", type = "source", repos = "https://h2o-release.s3.amazonaws.com/h2o/rel-weierstrass/2/R")

```
HOWEVER, I believe this may have contributed to the issue as I did this suggestion and it didn't help so I reverted back to version 3.14.0.7


Next, try to connect to Spark and run h2o_context() using as many stock settings as possible (i.e. get rid of all your conf and other settings). IF it works, then you know your configuration is somehow preventing h2o from booting (this is what was happening to me).

Finally I made changes to my configuration (followed the helpful guide here: https://stackoverflow.com/questions/37871194/how-to-tune-spark-executor-number-cores-and-executor-memory). I did notice in the Spark UI that I was never able to control the number of cores used (always at max) until I implemented the settings suggested in the guide. 

Here is what my config call looked like (48 core 256 GB RAM system):

```
config <- spark_config()
config$spark.executor.cores=5
config$spark.executor.memory="25g"
config$`sparklyr.shell.driver-memory` <- "6g"
config$`sparklyr.shell.executor-memory` <- "6g"
config$spark.executor.instances=9
config$spark.driver.memory="4g"
config$spark.dynamicAllocation.enabled='false'
```

For those of us using R and sparklyr be sure to set the sparklyr.shell.driver-memory and sparklyr.shell.executor-memory to above the default (1G) as h2o will not run properly at that memory setting.

Good Luck.

**Comment by jakubhava** [comment]
Thanks for the suggestion @chrismoran0124.

@navdeep-G this seems like bug in RSparkling ( we are suggesting older H2O version ). What do you think ?

```
Detected H2O version 3.14.0.7. Please install H2O version 3.14.0.2, which is compliant with the latest Sparkling Water version for Spark 2.2.* -> Sparkling Water version 2.2.2


                       To update your h2o R package, copy/paste the following commands and then restart your R session:
                       
                       detach("package:rsparkling", unload = TRUE)
                       if ("package:h2o" %in% search()) { detach("package:h2o", unload = TRUE) }
                       if (isNamespaceLoaded("h2o")){ unloadNamespace("h2o") }
                       remove.packages("h2o")
                       install.packages("h2o", type = "source", repos = "https://h2o-release.s3.amazonaws.com/h2o/rel-weierstrass/2/R")

```

**Comment by WeidongZhou** [comment]
@ibobak,
I figured it out the issue early this morning, also on a BDA. The issue we run into the error of "cloud size under xx" is because I try to run sparkling-shell on edge node. Once I installed the package on one of BDA node. It worked and I can see H2O Flow UI. From the sparking-shell, it shows that the UI address is the BDA node's InfinitBand private IP address. As edge node can not communicate to the BDA's private IB network, no wonder it is timed out and never get the response back. Hope it helps.

Weidong
